From 9dd7926dd6fd6fc078b99cc279c0c198b4178b65 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Wed, 29 Oct 2014 16:33:54 -0400
Subject: [fs] ext4: avoid unneeded lookup when xattr name is invalid

Message-id: <1414600485-1929-5-git-send-email-lczerner@redhat.com>
Patchwork-id: 98534
O-Subject: [RHEL 7.1 PATCH v2 04/55] ext4: avoid unneeded lookup when xattr name is invalid
Bugzilla: 1156096
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Zhang Zhen <zhenzhang.zhang@huawei.com>

BZ 1156096
https://bugzilla.redhat.com/show_bug.cgi?id=1156096

BREW 8169722
https://brewweb.devel.redhat.com/taskinfo?taskID=8169722

Upstream commit 230b8c1a7b388ac5e06549a7dbe083568875804e

 In ext4_xattr_set_handle() we have checked the xattr name's length. So
 we should also check it in ext4_xattr_get() to avoid unneeded lookup
 caused by invalid name.

 Signed-off-by: Zhang Zhen <zhenzhang.zhang@huawei.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/xattr.c b/fs/ext4/xattr.c
index 2e80c88..3961a52 100644
--- a/fs/ext4/xattr.c
+++ b/fs/ext4/xattr.c
@@ -367,6 +367,9 @@ ext4_xattr_get(struct inode *inode, int name_index, const char *name,
 {
  int error;
 
+ if (strlen(name) > 255)
+  return -ERANGE;
+
  down_read(&EXT4_I(inode)->xattr_sem);
  error = ext4_xattr_ibody_get(inode, name_index, name, buffer,
          buffer_size);
-- 
1.7.1